edge-modules/DirectMethodCloudSender/Settings.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace DirectMethodCloudSender
+{
+    using System;
+    using System.IO;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+    using Newtonsoft.Json.Serialization;
+
+    [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+    public class Settings
+    {
+        static readonly Lazy<Settings> DefaultSettings = new Lazy<Settings>(
+            () =>
+            {
+                IConfiguration configuration = new ConfigurationBuilder()
+                    .SetBasePath(Directory.GetCurrentDirectory())
+                    .AddJsonFile("config/settings.json", optional: true)
+                    .AddEnvironmentVariables()
+                    .Build();
+
+                return new Settings(
+                    configuration.GetValue<string>("IOTEDGE_DEVICEID"),
+                    configuration.GetValue<string>("TargetModuleId", "DirectMethodReceiver"),
+                    configuration.GetValue<TransportType>("TransportType", TransportType.Amqp_Tcp_Only),
+                    configuration.GetValue<TimeSpan>("DirectMethodDelay", TimeSpan.FromSeconds(5)),
+                    configuration.GetValue<Uri>("AnalyzerUrl", new Uri("http://analyzer:15000")),
+                    configuration.GetValue<string>("ServiceClientConnectionString"));
+            });
+
+        Settings(
+            string deviceId,
+            string targetModuleId,
+            TransportType transportType,
+            TimeSpan directMethodDelay,
+            Uri analyzerUrl,
+            string serviceClientConnectionString)
+        {
+            this.DeviceId = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.TargetModuleId = Preconditions.CheckNonWhiteSpace(targetModuleId, nameof(targetModuleId));
+            this.TransportType = Preconditions.CheckNotNull(transportType, nameof(transportType));
+            this.DirectMethodDelay = Preconditions.CheckNotNull(directMethodDelay);
```

Comment:
>CheckNotNull [](start = 51, length = 12)

it can't be null because TimeSpan is struct

Reply:
yes, please make sure you use the same patttern

---
In reply to: [354614192](https://github.com/Azure/iotedge/pull/2074#discussion_r354614192) [](ancestors = 354614192)